### PR TITLE
TRC/S: Avoid flooding with logs

### PR DESCRIPTION
**Pull request by shgandhi**

### CODE REVIEW

**Review by teoparvanov — APPROVED** [review]
LGTM!

PS: we should definitely consider integrating Concord Client with Apollo so that we test cover such scenarios before bumping into Concord 😉

**Comment by shgandhi** [comment]
> LGTM!
> 
> PS: we should definitely consider integrating Concord Client with Apollo so that we test cover such scenarios before bumping into Concord 😉

Thanks for the review @teoparvanov. Yes, if all goes well, we will start working on that soon.

**Review by robem — APPROVED** [review]

File: client/thin-replica-client/src/thin_replica_client.cpp [hunk 1]
```diff
@@ -541,6 +544,19 @@ void ThinReplicaClient::receiveUpdates() {
...
+        LOG_WARN(logger_, "Couldn't find agreement amongst all servers. Try again.");
+        last_non_agreement_time = std::chrono::steady_clock::now();
+      } else {
+        auto current_time = std::chrono::steady_clock::now();
+        auto duration =
+            std::chrono::duration_cast<std::chrono::seconds>(current_time - last_non_agreement_time.value());
+        if (duration.count() >= config_->duration.count()) {
+          LOG_WARN(logger_, "Couldn't find agreement amongst all servers. Try again.");
+        }
+      }
```
[the remark this answers is not in the document]
Comment (shgandhi) [reply client/thin-replica-client/src/thin_replica_client.cpp]: Updated.

File: client/thin-replica-client/src/thin_replica_client.cpp [hunk 2]
```diff
@@ -541,6 +544,19 @@ void ThinReplicaClient::receiveUpdates() {
 
     // At this point we need to have agreeing servers.
     if (most_agreeing < (config_->max_faulty + 1)) {
+      // print the warning every minute to avoid flooding with logs
+      if (!last_non_agreement_time.has_value()) {
+        LOG_WARN(logger_, "Couldn't find agreement amongst all servers. Try again.");
+        last_non_agreement_time = std::chrono::steady_clock::now();
+      } else {
+        auto current_time = std::chrono::steady_clock::now();
+        auto duration =
```
[the remark this answers is not in the document]
Comment (shgandhi) [reply client/thin-replica-client/src/thin_replica_client.cpp]: Updated.

File: client/thin-replica-client/include/client/thin-replica-client/thin_replica_client.hpp [hunk 3]
```diff
@@ -224,15 +224,23 @@ struct ThinReplicaClientConfig {
   // trs_conns is a vector of connection objects. Each representing a direct
   // connection from this TRC to a specific Thin Replica Server.
   std::vector<std::unique_ptr<TrsConnection>> trs_conns;
+  // the time duration the TRC waits before printing warning logs when
+  // responsive agreeing servers are less than config_->max_faulty + 1
+  std::chrono::seconds duration;
```
[the remark this answers is not in the document]
Comment (shgandhi) [reply client/thin-replica-client/include/client/thin-replica-client/thin_replica_client.hpp]: Yep, updated.

**Comment by shgandhi** [comment]
Thanks for the review @robem, updated the PR according to your comments.

**Pull request merged**